experimental/segmenter/src/provider/lstm.rs

```diff
@@ -0,0 +1,362 @@
+
+        use core::ops::Deref;
+        use serde::de::Error;
+        Self::from_parts(
+            raw.dims
+                .deref()
+                .try_into()
+                .map_err(|_| S::Error::custom("Dimension mismatch"))?,
+            raw.data,
+        )
+        .map_err(|_| S::Error::custom("Dimension mismatch"))
+    }
+}
+
+#[cfg(feature = "datagen")]
+impl<const D: usize> serde::Serialize for LstmMatrix<'_, D> {
+    fn serialize<S: serde::ser::Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
+        use serde::ser::SerializeStruct;
+        let mut state = serializer.serialize_struct("LstmMatrix", 2)?;
+        state.serialize_field("dims", self.dims.as_slice())?;
+        state.serialize_field("data", &self.data)?;
+        state.end()
+    }
+}
+
+#[cfg(feature = "datagen")]
+impl<const D: usize> databake::Bake for LstmMatrix<'_, D> {
+    fn bake(&self, env: &databake::CrateEnv) -> databake::TokenStream {
+        let dims = self.dims.bake(env);
+        let data = self.data.bake(env);
+        databake::quote! {
+            icu_segmenter::provider::LstmMatrix::from_parts_unchecked(#dims, #data)
+        }
+    }
+}
+
+#[derive(PartialEq, Debug, Clone, Copy)]
+#[cfg_attr(
+    feature = "datagen",
+    derive(serde::Serialize,databake::Bake),
+    databake(path = icu_segmenter::provider),
+)]
+#[cfg_attr(feature = "serde", derive(serde::Deserialize))]
+/// The type of LSTM model
+pub enum ModelType {
+    /// A model working on code points
+    Codepoints,
+    /// A model working on grapheme clusters
+    GraphemeClusters,
+}
+
+/// The struct that stores a LSTM model.
+#[icu_provider::data_struct(LstmDataV1Marker = "segmenter/lstm@1")]
+#[derive(PartialEq, Debug, Clone)]
+#[cfg_attr(feature = "datagen", derive(serde::Serialize))]
+#[yoke(prove_covariance_manually)]
+pub struct LstmDataV1<'data> {
+    /// Type of the model
+    pub(crate) model: ModelType,
+    /// The grapheme cluster dictionary used to train the model
+    pub(crate) dic: ZeroMap<'data, str, u16>,
+    /// The embedding layer
+    pub(crate) embedding: LstmMatrix<'data, 2>,
+    /// The forward layer's first matrix
+    pub(crate) fw_w: LstmMatrix<'data, 3>,
+    /// The forward layer's second matrix
+    pub(crate) fw_u: LstmMatrix<'data, 3>,
+    /// The forward layer's bias
+    pub(crate) fw_b: LstmMatrix<'data, 2>,
+    /// The backward layer's first matrix
+    pub(crate) bw_w: LstmMatrix<'data, 3>,
+    /// The backward layer's second matrix
+    pub(crate) bw_u: LstmMatrix<'data, 3>,
+    /// The backward layer's bias
+    pub(crate) bw_b: LstmMatrix<'data, 2>,
+    /// The output layer's weights
+    pub(crate) time_w: LstmMatrix<'data, 3>,
+    /// The output layer's bias
+    pub(crate) time_b: LstmMatrix<'data, 1>,
+}
+
+impl<'data> LstmDataV1<'data> {
+    #[doc(hidden)] // databake
+    #[allow(clippy::too_many_arguments)] // constructor
+    pub const fn from_parts_unchecked(
```

Comment:
Nit: Should this be `unsafe` since there are invariants at play?

Since https://github.com/unicode-org/icu4x/commit/c4af14077a33191e4d739f0486aaf6d47a54bced, the `Lstm` type has hard invariants that must be enforced.